File: templates/python/calibration.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""校准曲线 —— 卡片 06_预测模型评价图/校准曲线_Calibration.md
合成预测 + sklearn calibration_curve。改 # >>> PARAM 区换 y_true / y_prob。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, matplotlib.pyplot as plt
from sklearn.calibration import calibration_curve
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_calibration"
TITLE = "Calibration curve"; XLAB = "Predicted probability"; YLAB = "Observed frequency"
CURVE_COLOR = pal("med_case_control")["low"]
N_BINS = 10
FIGSIZE = (3.2, 3.2)
# 换真实数据： y_true=df["y"]; y_prob=model.predict_proba(X)[:,1]
n = 2000
y_prob = np.random.beta(2, 3, n)
y_true = np.random.binomial(1, y_prob)
# <<< PARAM ------------------------------------------------------

pt, pp = calibration_curve(y_true, y_prob, n_bins=N_BINS, strategy="quantile")
brier = np.mean((y_prob - y_true) ** 2)
fig, ax = plt.subplots(figsize=FIGSIZE)
ax.plot([0, 1], [0, 1], ls="--", c="grey", lw=.8)
ax.plot(pp, pt, "o-", color=CURVE_COLOR)
ax.plot(y_prob, np.full(n, -.02), "|", color="k", alpha=.03)
ax.text(.05, .92, f"Brier = {brier:.3f}", transform=ax.transAxes, fontsize=9)
ax.set_xlim(-.03, 1); ax.set_ylim(-.05, 1); ax.set_aspect("equal"); ax.grid(False)
ax.set_xlabel(XLAB); ax.set_ylabel(YLAB); ax.set_title(TITLE)
save_fig(fig, NAME)

File: templates/r/calibration.R
## 校准曲线（loess + 分箱点 + Brier）—— 卡片 06_预测模型评价图/校准曲线_Calibration.md
## 合成预测，可独立运行： Rscript templates/r/calibration.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_calibration"; TITLE <- "Calibration curve"
XLAB <- "Predicted probability"; YLAB <- "Observed frequency"
CURVE <- unname(pal("med_case_control")["low"]); PTS <- unname(pal("med_case_control")["high"])
# 换真实数据： y(0/1) 与 pred(预测概率)
n <- 2000; pred <- rbeta(n, 2, 3); y <- rbinom(n, 1, pred)
# <<< PARAM ------------------------------------------------------

dat <- data.frame(pred = pred, y = y)
brier <- mean((dat$pred - dat$y)^2)
qbin <- cut(dat$pred, breaks = quantile(dat$pred, 0:10 / 10), include.lowest = TRUE)
binpts <- data.frame(pred = tapply(dat$pred, qbin, mean), obs = tapply(dat$y, qbin, mean))
p <- ggplot(dat, aes(pred, y)) +
  geom_abline(slope = 1, intercept = 0, linetype = 2, color = "grey60") +
  geom_smooth(method = "loess", se = TRUE, color = CURVE, linewidth = .8) +
  geom_point(data = binpts, aes(pred, obs), color = PTS, size = 1.8, inherit.aes = FALSE) +
  geom_rug(sides = "b", alpha = .05) +
  annotate("text", x = .05, y = .92, hjust = 0, size = 2.8, label = sprintf("Brier = %.3f", brier)) +
  coord_equal(xlim = c(0, 1), ylim = c(0, 1)) +
  labs(x = XLAB, y = YLAB, title = TITLE)
save_gg(p, NAME, w = 88, h = 88)
